### PR TITLE
home work #17

### PR DESCRIPTION
**Pull request by DeliriumInTheVoid**

### CODE REVIEW

File: lesson_17/headers/custom_string.hpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+{
+    delete[] m_str; m_str = nullptr;// MEMORY LEAK
+    if (c_str == nullptr)
+    {
+        m_str = new char[1]; //m_str = new char(?);
+        *m_str = '\0';
+    }
+    else
+    {
+        const std::size_t size = strlen(c_str);
```
Comment (helen-black) [inline lesson_17/headers/custom_string.hpp:39]: it should be `strlen(c_str) + 1;` tto reserve place for '\0' symbol

File: lesson_17/headers/custom_string.hpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+int* allocateArray(int size)
+{
+    int* arrayInt = new int[size];
+    return arrayInt;
+}
+
+void clearArray(int* numbers)
+{
+    delete numbers;
```
Comment (helen-black) [inline lesson_17/headers/custom_string.hpp:67]: memory leak here

File: lesson_17/sources/dynamic_int_array.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+#include "../headers/dynamic_int_array.hpp"
+
+
+//default c-tor
+dynamic_int_array::dynamic_int_array() : dynamic_int_array{ 0 }
+{
+}
+
+//c-tor with size
+dynamic_int_array::dynamic_int_array(const std::size_t size) : size_{ 0 }
```
Comment (helen-black) [inline lesson_17/sources/dynamic_int_array.cpp:13]: wrong size initialization

File: lesson_17/sources/dynamic_int_array.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    if (this != &other)
+    {
+        delete[] dynamic_buffer_;
+        dynamic_buffer_ = nullptr;
+
+        max_size_ = other.max_size_;
+        size_ = other.size_;
+
+        dynamic_buffer_ = new int[max_size_];
+        std::copy_n(other.dynamic_buffer_, size_, dynamic_buffer_);
```
Comment (helen-black) [inline lesson_17/sources/dynamic_int_array.cpp:46]: the same as copy c-tor

File: lesson_17/sources/dynamic_int_array.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+//delete previous memory, create new inner array with updated size
+void dynamic_int_array::set_size(const std::size_t new_size)
+{
+    delete[] dynamic_buffer_;
+    dynamic_buffer_ = nullptr;
+
+    size_ = 0;
+    max_size_ = new_size > initial_buffer_size ? new_size : initial_buffer_size;
+    dynamic_buffer_ = new int[max_size_];
+}
```
Comment (helen-black) [inline lesson_17/sources/dynamic_int_array.cpp:71]: You should store data that was in array. If new size is less then previous crop the array, if it is bigger add zeros

**Review by helen-black — CHANGES_REQUESTED** [review]

File: lesson_17/headers/custom_string.hpp (hunk 1 above)
Comment (DeliriumInTheVoid) [reply lesson_17/headers/custom_string.hpp]: done

File: lesson_17/headers/custom_string.hpp (hunk 2 above)
Comment (DeliriumInTheVoid) [reply lesson_17/headers/custom_string.hpp]: done

File: lesson_17/sources/dynamic_int_array.cpp (hunk 5 above)
Comment (DeliriumInTheVoid) [reply lesson_17/sources/dynamic_int_array.cpp]: done

**Review by helen-black — APPROVED** [review]

**Pull request merged**